Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp

```diff
@@ -713,217 +716,165 @@ class CombinatorialKalmanFilter {
-              stepper.update(state.stepping,
-                             MultiTrajectoryHelpers::freeFiltered(
-                                 state.options.geoContext, currentState),
-                             currentState.filtered(),
-                             currentState.filteredCovariance(), *surface);
-              ACTS_VERBOSE(
-                  "Stepping state is updated with filtered parameter:");
-              ACTS_VERBOSE("-> " << currentState.filtered().transpose()
-                                 << " of track state with tip = "
-                                 << currentState.index());
-            }
-          }
+          return Result<void>::success();
         }
 
-        // Update state and stepper with post material effects
-        materialInteractor(surface, state, stepper, navigator,
-                           MaterialUpdateStage::PostUpdate);
-      } else if (isSensitive || isMaterial) {
-        ACTS_VERBOSE("Handle " << (isSensitive ? "sensitive" : "passive")
-                               << " surface: " << surface->geometryId()
-                               << " without measurements");
-
-        // No splitting on the surface without source links. Set it to one
-        // first, but could be changed later
-        nBranchesOnSurface = 1;
-
-        auto currentBranch = result.activeBranches.back();
-        TrackIndexType prevTip = currentBranch.tipIndex();
+        // `currentBranch` is invalidated after `processNewTrackStates`
+        currentBranch = result.activeBranches.back();
+        prevTip = currentBranch.tipIndex();
+      } else {
+        if (expectMeasurements) {
+          ACTS_VERBOSE("Detected hole after measurement selection on surface "
+                       << surface->geometryId());
+        }
 
-        // No source links on surface, add either hole or passive material
-        // TrackState. No storage allocation for uncalibrated/calibrated
-        // measurement and filtered parameter
         auto stateMask = PM::Predicted | PM::Jacobian;
 
-        // Transport the covariance to a curvilinear surface
-        stepper.transportCovarianceToCurvilinear(state.stepping);
-
-        // Update state and stepper with pre material effects
-        materialInteractor(surface, state, stepper, navigator,
-                           MaterialUpdateStage::PreUpdate);
-
-        // Transport & bind the state to the current surface
-        auto boundStateRes =
-            stepper.boundState(state.stepping, *surface, false);
-        if (!boundStateRes.ok()) {
-          return boundStateRes.error();
-        }
-        auto& boundState = *boundStateRes;
-        auto& [boundParams, jacobian, pathLength] = boundState;
-        boundParams.covariance() = state.stepping.cov;
-
-        // Add a hole or material track state to the multitrajectory
+        // Add a hole track state to the multitrajectory
         TrackIndexType currentTip = addNonSourcelinkState(
```

Comment:
I don't see any other call to `addNonSourcelinkState` now. Are material-only surfaces not recorded as track states anymore? Don't we also get here if `expectMeasurements == false`, which case it's a material state? If so I suppose the command above is incorrect?